src/com/amazon/ion/impl/IonReaderBinaryIncremental.java

```diff
@@ -0,0 +1,1932 @@
+    private static final int MAX_ID_ID = 8;
+
+    // The size of the reusable UTF-8 decoding buffer.
+    private static final int UTF8_BUFFER_SIZE_IN_BYTES = 4 * 1024;
+
+    // The final byte of the binary IVM.
+    private static final int IVM_FINAL_BYTE = 0xEA;
+
+    // Isolates the highest bit in a byte.
+    private static final int HIGHEST_BIT_BITMASK = 0x80;
+
+    // Isolates the lowest seven bits in a byte.
+    private static final int LOWER_SEVEN_BITS_BITMASK = 0x7F;
+
+    // Isolates the lowest six bits in a byte.
+    private static final int LOWER_SIX_BITS_BITMASK = 0x3F;
+
+    // The number of significant bits in each UInt byte.
+    private static final int VALUE_BITS_PER_UINT_BYTE = 8;
+
+    // The number of significant bits in each VarUInt byte.
+    private static final int VALUE_BITS_PER_VARUINT_BYTE = 7;
+
+    // An IonCatalog containing zero shared symbol tables.
+    private static final IonCatalog EMPTY_CATALOG = new SimpleCatalog();
+
+    // Initial capacity of the stack used to hold ContainerInfo. Each additional level of nesting in the data requires
+    // a new ContainerInfo. Depths greater than 8 will be rare.
+    private static final int CONTAINER_STACK_INITIAL_CAPACITY = 8;
+
+    // Initial capacity of the ArrayList used to hold the symbol IDs of the annotations on the current value.
+    private static final int ANNOTATIONS_LIST_INITIAL_CAPACITY = 8;
+
+    // Initial capacity of the ArrayList used to hold the text in the current symbol table.
+    private static final int SYMBOLS_LIST_INITIAL_CAPACITY = 128;
+
+    // Single byte negative zero, represented as a VarInt. Often used in timestamp encodings to indicate unknown local
+    // offset.
+    private static final int VAR_INT_NEGATIVE_ZERO = 0xC0;
+
+    // The number of bytes occupied by a Java int.
+    private static final int INT_SIZE_IN_BYTES = 4;
+
+    // The number of bytes occupied by a Java long.
+    private static final int LONG_SIZE_IN_BYTES = 8;
+
+    // The smallest negative 8-byte integer that can fit in a long is -0x80_00_00_00_00_00_00_00.
+    private static final int MOST_SIGNIFICANT_BYTE_OF_MIN_LONG = 0x80;
+
+    // The largest positive 8-byte integer that can fit in a long is 0x7F_FF_FF_FF_FF_FF_FF_FF.
+    private static final int MOST_SIGNIFICANT_BYTE_OF_MAX_LONG = 0x7F;
+
+    // The second-most significant bit in the most significant byte of a VarInt is the sign.
+    private static final int VAR_INT_SIGN_BITMASK = 0x40;
+
+    // 32-bit floats must declare length 4.
+    private static final int FLOAT_32_BYTE_LENGTH = 4;
+
+    // The InputStream that provides the binary Ion data.
+    private final InputStream inputStream;
+
+    // Wrapper for the InputStream that ensures an entire top-level value is available.
+    private final IonReaderLookaheadBuffer lookahead;
+
+    // Buffer that stores top-level values.
+    private final ResizingPipedInputStream buffer;
```

Comment:
`buffer` is stored as a field in `IonReaderBinaryIncremental` just to avoid having to call `lookahead.getPipe()` in 22 different places. It may be personal preference.

Would you be more comfortable with adding `peek`, `getByteBuffer`, `seekTo`, and `copyBytes` methods on `IonReaderLookaheadBuffer` that call through to the corresponding methods on its `ResizingPipedInputStream`? Then removing the `buffer` field in `IonReaderBinaryIncremental` and only calling `IonReaderLookaheadBuffer` methods.

Reply:
I don't know that that's the right solution, either. My _feeling_  is that this pattern of abstraction penetration means that we don't have the right abstractions here, but like I said earlier I don't have a positive recommendation for how to factor this. Maybe I'm echoing misgivings you already had, maybe you're comfortable with it as it lies. There's no need to make a change to fix a symptom (taking/accessing a field from another class) only technically.

In any case none of this backs us into a corner, and I'm fine shipping this as-is. Should we find a better factoring we can change it then.

Reply:
First I'll share a little bit of background. The `IonReaderLookaheadBuffer` can be used in two ways:
1. It is instantiated directly by a user, who provides it a potentially-growing InputStream. The user interacts only with the public interface, e.g. `fillInput()` and `moreDataRequired()`. The user may construct an IonReader via `newIonReader()`; if this is a non-incremental reader, it is the user's responsibility to make sure that `next()` is called on the reader only after `moreDataRequired()` returns `false`. It was initially designed and used for this purpose.
2. It is instantiated internally by the incremental reader, and is used as the reader's internal buffer. So that it could continue being used in the way described in case 1, the buffer was exposed via a protected getter, rather than overhauling the abstraction.

For case 2, would you prefer that the incremental reader to instantiate its own buffer and provide it to the IonReaderLookaheadBuffer?